sources/imdsclient/src/lib.rs

```diff
@@ -245,7 +244,13 @@ impl ImdsClient {
         timeout(
             self.retry_timeout,
             Retry::spawn(retry_strategy(), || async {
-                let session_token = self.read_token().await?;
+                let session_token = match self.read_token().await? {
+                    Some(session_token) => session_token,
+                    None => {
+                        self.write_token().await?;
+                        error::EmptyToken.fail()?
+                    }
+                };
```

Comment:
Since this triggers a retry we are adding about 500ms to the first call of each new ImdsClient, but it would not impact subsequent calls. I'm not sure this is noticeable, but if it's a concern we can change the backoff strategy to start with 250ms or maybe add a nested read before error.

Reply:
Maybe `write_token` can return a clone of the fetched token, that way if it succeeds here we don't need to trigger a retry?